### PR TITLE
feat(resources/export): export deployed security groups as keel crds

### PR DESCRIPTION
**Pull request by asher**

### CODE REVIEW

File: keel-clouddriver/src/main/kotlin/com/netflix/spinnaker/keel/clouddriver/MemoryCloudDriverCache.kt [hunk 1]
```diff
@@ -60,7 +60,7 @@ class MemoryCloudDriverCache(
     .expireAfterWrite(1, HOURS)
     .build<String, Subnet>()
 
-  private fun credentialBy(name: String): Credential =
+  override fun credentialBy(name: String): Credential =
```
Comment (asher) [inline keel-clouddriver/src/main/kotlin/com/netflix/spinnaker/keel/clouddriver/MemoryCloudDriverCache.kt]: Making this a public part of the `CloudDriverCache` interface to assist with determining the locations configured with an account.

File: keel-plugin/src/main/kotlin/com/netflix/spinnaker/keel/plugin/ResourceHandler.kt [hunk 2]
```diff
@@ -156,6 +157,15 @@ abstract class ResourceHandler<S : ResourceSpec, R : Any>(
    */
   open suspend fun delete(resource: Resource<S>): List<Task> = TODO("Not implemented")
 
+  /**
+   * Generate a Keel SubmittedResource CRD from currently existing resources.
```
Comment (emjburns) [inline keel-plugin/src/main/kotlin/com/netflix/spinnaker/keel/plugin/ResourceHandler.kt]: Can we say `definition` here instead of CRD?

File: keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resource/SecurityGroupHandler.kt [hunk 3]
```diff
@@ -134,11 +140,111 @@ class SecurityGroupHandler(
...
+      } catch (e: ResourceNotFound) {
+        null
+      }
+    }
+      .filterValues { it != null }
+
+    val securityGroups =
+      coroutineScope {
+        summaries.map { (region, summary) ->
+          async {
```
Comment (emjburns) [inline keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resource/SecurityGroupHandler.kt]: If you're using `async` here, do you want to wait for everything to complete at all?

Comment (emjburns) [reply keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resource/SecurityGroupHandler.kt]: Ah I see the `await` at the end of this, never mind. I'm still new to these fancy things.

File: keel-api/src/main/kotlin/com/netflix/spinnaker/keel/rest/ExportController.kt [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+    produces = [MediaType.APPLICATION_JSON_VALUE, APPLICATION_YAML_VALUE]
+  )
+  fun get(
+    @PathVariable("cloudProvider") cloudProvider: String,
+    @PathVariable("account") account: String,
+    @PathVariable("type") type: String,
+    @PathVariable("name") name: String,
+    @RequestParam("serviceAccount") serviceAccount: String
+  ): SubmittedResource<*> {
+    val apiVersion = SPINNAKER_API_V1.subApi(cloudProviderOverrides[cloudProvider] ?: cloudProvider)
```
Comment (robfletcher) [inline keel-api/src/main/kotlin/com/netflix/spinnaker/keel/rest/ExportController.kt]: I’m wondering if we should have a way of registering these API versions

File: keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resource/SecurityGroupHandler.kt [hunk 5]
```diff
@@ -134,11 +140,111 @@ class SecurityGroupHandler(
...
+      )
+    )
+  }
+
   private fun ResourceDiff<Map<String, SecurityGroup>>.toIndividualDiffs() =
     desired.map { (region, desire) ->
       ResourceDiff(desire, current?.getOrDefault(region, null))
     }
 
+  private fun SecurityGroupSpec.generateOverrides(
```
Comment (emjburns) [inline keel-ec2-plugin/src/main/kotlin/com/netflix/spinnaker/keel/ec2/resource/SecurityGroupHandler.kt]: oh nice!

File: keel-api/src/main/kotlin/com/netflix/spinnaker/keel/rest/ExportController.kt [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+    val kind = CaseFormat.LOWER_CAMEL.to(CaseFormat.LOWER_HYPHEN, type)
```
Comment (robfletcher) [inline keel-api/src/main/kotlin/com/netflix/spinnaker/keel/rest/ExportController.kt]: Why do we not just expect the kebab case in the URL?

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/api/Exportable.kt [hunk 7]
```diff
@@ -0,0 +1,11 @@
+package com.netflix.spinnaker.keel.api
+
+import com.netflix.spinnaker.keel.model.Moniker
+
+data class Exportable(
+  val account: String,
```
Comment (robfletcher) [inline keel-core/src/main/kotlin/com/netflix/spinnaker/keel/api/Exportable.kt]: Is `account` an EC2-specific concept? Or at least one that may not be common to all cloud providers?

File: keel-plugin/src/main/kotlin/com/netflix/spinnaker/keel/plugin/ResourceHandler.kt [hunk 8]
```diff
@@ -173,7 +183,10 @@ fun Collection<ResourceHandler<*, *>>.supporting(
   apiVersion: ApiVersion,
   kind: String
 ): ResourceHandler<*, *> =
-  find { it.apiVersion == apiVersion && it.supportedKind.first.singular == kind }
+  find {
+    it.apiVersion == apiVersion &&
+      (it.supportedKind.first.singular == kind || it.supportedKind.first.plural == kind)
```
Comment (robfletcher) [inline keel-plugin/src/main/kotlin/com/netflix/spinnaker/keel/plugin/ResourceHandler.kt]: Curious why this change is necessary

Comment (asher) [reply keel-plugin/src/main/kotlin/com/netflix/spinnaker/keel/plugin/ResourceHandler.kt]: Clouddriver types are generally plural but I don't know if this is true for all cloud providers.

File: keel-core/src/main/kotlin/com/netflix/spinnaker/keel/api/Exportable.kt (hunk 7 above)
Comment (asher) [reply keel-core/src/main/kotlin/com/netflix/spinnaker/keel/api/Exportable.kt]: No, its lower level and should be required by all cloudProviders. If you deploy to 3 k8s clusters, each cluster should appear as a unique account.

File: keel-api/src/main/kotlin/com/netflix/spinnaker/keel/rest/ExportController.kt (hunk 6 above)
Comment (asher) [reply keel-api/src/main/kotlin/com/netflix/spinnaker/keel/rest/ExportController.kt]: This REST method attempts to be compatible with how the rest of Spinnaker views resources. We could expect deck or gate to convert `securityGroups` to `security-group` (and forgo the plural kind match you asked about below) or `security-groups` but I don't know if that sort of conversion will always be correct or if we should burden callers with knowledge of how clouddriver types map to keel kinds.

File: keel-api/src/main/kotlin/com/netflix/spinnaker/keel/rest/ExportController.kt (hunk 4 above)
Comment (asher) [reply keel-api/src/main/kotlin/com/netflix/spinnaker/keel/rest/ExportController.kt]: +1 (though not currently sure what it'd look like)
